### DISCUSSION TITLE
Events dispatched from Mailer

### DISCUSSION THREAD

**Discussion by ju5t** [laravel/framework#42245] [Q&A]
@driesvints sorry for barging into your notifications like this, but because you made the PR for Symfony's mailer, I'm trying my luck here. Sorry for that!

We have an event listener on `Illuminate\Mail\Events\MessageSent`. We have a requirement to store outgoing emails and link them to a related model. The `MessageSent` event only contains a message along with a data array. We enhance the data array with `buildViewDataUsing()` in the `Mailable`, which gives us the data we need in our listener. All well so far. It works.

Testing this beast is different. Our test looks like this:

```php
Mail::fake();
Event::fake();

Mail::to('foo@bar.com')
    ->send(new AwesomeEmail());

Event::assertDispatched(MessageSent::class, function ($event) {
    $this->assertEquals($event->data['__private'], 'private');
});
```

This does't catch the `MessageSent` event, neither did it when removing `Mail::fake()`. I did some source surfing, and saw that `Mailer` accepts `Dispatcher` as an argument when constructing the class. The `MailManager` is doing this, it fetches `events` from the IoC and passes it on.

I get all of this. 

But I can't wrap my head around why `Event::fake()` isn't seeing the `MessageSent` event. I have guesses:

- It's because the `mailer` is initialised before I run `Event::fake()` and as such it can't catch the event? I'm not sure.
- It's because our tests use the `ArrayTransport` which doesn't send an event? Removing `MAIL_MAILER` from my phpunit.xml has no effect.

The Framework tests for `MessageSent` mocks the dispatcher and creates its own `Mailer`. I'm afraid that if I do that, I end up mocking too much; and it's not necessarily something I prefer.

How can I make my test work?

**Comment by driesvints** [comment]
Hmm, I'll investigate this. Can you create a repo? 

Heya, thanks for reporting.

We'll need more info and/or code to debug this further. Can you please create a repository with the command below, commit the code that reproduces the issue as separate commits on the main/master branch and share the repository here? Please make sure that you have the [latest version of the Laravel installer](https://github.com/laravel/installer) in order to run this command. Please also make sure you have both Git & [the GitHub CLI tool](https://cli.github.com) properly set up.

```bash
laravel new bug-report --github="--public"
```

Please do not amend and create a separate commit with your custom changes. After you've posted the repository, we'll try to reproduce the issue.

Thanks!

**Reply by ju5t** [reply]
First time for me I'm raising a bug report here so sorry if things are a bit clumsy on my end.

https://github.com/ju5t/bug-report/commit/8650fd6d248c85469c40976bd4f9ed301808f4c6

In our own app we'd like to assert the event as well by passing a closure, but that's out of scope for now I guess :)

**Comment by driesvints** [comment]
Ah sorry @ju5t. I just realized this indeed won't work. You can't assert that the MessageSent event was fired because it doesn't get fired with the MailFake class.

**Reply by ju5t** [reply]
No problem. It's weird why it doesn't work for me without `Mail::fake`, but that must be something app specific. In the bug-report repo all is well if you remove `Mail::fake`. We're using the ArrayTransport in our tests so I assume `Mail::fake` is only really useful when you want to assert things?

Would there be another way to test the event without mocking everything?

**Reply by driesvints** [reply]
> We're using the ArrayTransport in our tests so I assume Mail::fake is only really useful when you want to assert things?

yes indeed.

> Would there be another way to test the event without mocking everything?

I don't think it makes sense to combine this with mail mocking. You already can assert the mail was sent so you can be sure the event will be sent as well.

**Reply by ju5t** [reply]
I have no doubts about the event itself. We're using `buildViewDataUsing` to pass additional data to the view. Coincidentally, this data is also available in the `MessageSent` event, which is where we fetch the data from in our e-mail logger. The only thing I want to know is if that data is available in the event.

I found that the below doesn't work by the way.

```php
Event::assertDispatched(MessageSent::class, function ($event) {
    $this->assertSame('foo', $event->data['__private']);
});
```

But this does:

```php
Event::assertDispatched(MessageSent::class);
```

(in our own app we appear to have a completely different issue for some reason)

**Reply by ju5t** [reply]
I found something weird in our app. I understand that this is something incredibly hard to help with, but I hope you can give me some directions where to look next. If there is paid support somewhere or if you know someone who can help with this, there is budget for it!

In `Illuminate\Mail\Mailer` there is `dispatchSentEvent`:

```php
protected function dispatchSentEvent($message, $data = [])
{
    if ($this->events) {
        $this->events->dispatch(
            new MessageSent($message, $data)
        );
    }
}
```

As I said, for some weird reason it doesn't emit the event when I use `Event::fake`. When I replace `$this-events->dispatch(..` for `        Event::dispatch(new MessageSent($message, $data));`, it works. This problem isn't there in a new Laravel app so it's 100% on us.

In `dispatchSentEvent` I added some debug statements:

```php
$events = app('events');

ray(get_class($events));
ray(get_class($this->events));
```

Turns out the IoC returns `Illuminate\Support\Testing\Fakes\EventFake` as expected, but `$this->events` returns `Illuminate\Events\Dispatcher`. Which is probably why it doesn't work for us.

I know `$this->events` is set in the `MailManager`, what I don't get is why it's not being replaced by our faked instance. My best bet would be the order in which things are running but I have no clue.

I'm guessing this is why some people don't write tests, it can be pretty difficult at times :)

**Reply by BobbyBorisov** [reply]
hey @ju5t I am experiencing the same issue when upgrading from laravel 8 to laravel 9 my tests suddenly started to fail. Before upgrading `$this->events` was returning the `EventFake`, now on laravel 9 its returning `Dispatcher` and I am wondering how to test my event. Did you manage to fix it somehow? Cheers

**Reply by ju5t** [reply]
hi @BobbyBorisov, I don't believe we did. I checked our app, but it lacks a test for the behaviour we needed it for. It requires a manual test which we documented in our test procedure (yea not great).
